### PR TITLE
core: refactor GracefulSwitchLoadBalancer to switch based on LoadBalancer.Factory

### PR DESCRIPTION
**Pull request by dapengzhang0**
`GracefulSwitchLoadBalancer` was doing switch based on `LoadBalancerProvider.getPolicyName()`. This turned out to be very awkward when I have to synthesize a policy name for the provider, and what I actually care about is the identity of the lb provider not necessarily the policy name.

Now `GracefulSwitchLoadBalancer` is doing switch based on identity of `LoadBalancer.Factory`, which is simpler.

### CODE REVIEW

File: api/src/main/java/io/grpc/LoadBalancerProvider.java [hunk 1]
```diff
@@ -86,15 +86,16 @@ public final String toString() {
   }
 
   /**
-   * Uses identity equality.
+   * Equal provider instances must be of the same class, have the same policy name, priority and
```
Comment (voidzcy) [inline api/src/main/java/io/grpc/LoadBalancerProvider.java]: It's not appropriate to mention details that only applies to subclasses here (e.g., not all load balancers have lb configs).

Comment (dapengzhang0) [reply api/src/main/java/io/grpc/LoadBalancerProvider.java]: Here it's the provider not balancer that parses lb config. 

``` java
public ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> rawLoadBalancingPolicyConfig)
```

The config could be empty or not relevant to the balancer at all.

Comment (voidzcy) [reply api/src/main/java/io/grpc/LoadBalancerProvider.java]: Well, I just realized that `parseLoadBalancingPolicyConfig(...)` is an instance method for `LoadBalancerProvider` (even for `PickFirstLoadBalancerProvider` it provides an implementation that returns a `NO_CONFIG`).

Well. I would leaning for not saying anything here. And rather, implementors should know what they are doing.

Comment (dapengzhang0) [reply api/src/main/java/io/grpc/LoadBalancerProvider.java]: I just noticed that `LoadBalancerProvider` has a much simpler super class `LoadBalancer.Factory`, and that is all I need now.

File: core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java [hunk 2]
```diff
@@ -77,21 +78,23 @@ public GracefulSwitchLoadBalancer(Helper helper) {
     this.helper = checkNotNull(helper, "helper");
   }
 
-  /** Gracefully switch to a new load balancing policy. */
+  /**
+   * Gracefully switch to a new balancer type. Two {@link LoadBalancerProvider}s are considered of
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: Done.

File: core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java [hunk 3]
```diff
@@ -60,13 +60,13 @@ public String toString() {
 
   private final Helper helper;
 
-  // While the new policy is not fully switched on, the pendingLb is handling new updates from name
+  // While the new factory is not fully switched on, the pendingLb is handling new updates from name
```
Comment (voidzcy) [inline core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: I feel like "policy" sounds better.

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: I agree. I changed just be cause it was literally factory. Now I define policy by factory.

File: core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java [hunk 4]
```diff
@@ -77,21 +76,23 @@ public GracefulSwitchLoadBalancer(Helper helper) {
     this.helper = checkNotNull(helper, "helper");
   }
 
-  /** Gracefully switch to a new load balancing policy. */
-  public void switchTo(LoadBalancerProvider newLbProvider) {
-    checkNotNull(newLbProvider, "newLbProvider");
+  /**
+   * Gracefully switch to a new policy defined by the given factory. Two factories are considered to
```
Comment (voidzcy) [inline core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: I don't think the second sentence is necessary. It only adds confusion. Other than that, LGTM.

Comment (zhangkun83) [reply core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: Maybe change it to "Gracefully switch to a new policy defined by the given factory, if the given factory isn't equal to the current one".

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java]: > if the given factory isn't equal to the current one".

hmm, sounds better.

**Pull request merged**